src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/net/NetClient.java

```diff
@@ -83,27 +103,36 @@ public void flushStream(final String remoteHost) {
   }
 
   private void closeAllDataStreams() {
-    for (Map.Entry<String, StreamObserver<FlowUnitMessage>> entry :
+    for (Map.Entry<String, AtomicReference<StreamObserver<FlowUnitMessage>>> entry :
         perHostOpenDataStreamMap.entrySet()) {
       LOG.debug("Closing stream for host: {}", entry.getKey());
       // Sending an onCompleted should trigger the subscriber's node state manager
       // and cause this host to be put under observation.f
-      entry.getValue().onCompleted();
+      entry.getValue().get().onCompleted();
       perHostOpenDataStreamMap.remove(entry.getKey());
     }
   }
 
   private StreamObserver<FlowUnitMessage> getDataStreamForHost(
       final String remoteHost, final StreamObserver<PublishResponse> serverResponseStream) {
     if (perHostOpenDataStreamMap.containsKey(remoteHost)) {
-      return perHostOpenDataStreamMap.get(remoteHost);
+      return perHostOpenDataStreamMap.get(remoteHost).get();
     }
+    return addOrUpdateDataStreamForHost(remoteHost, serverResponseStream);
+  }
 
+  private synchronized StreamObserver<FlowUnitMessage> addOrUpdateDataStreamForHost(
+      final String remoteHost, final StreamObserver<PublishResponse> serverResponseStream) {
     InterNodeRpcServiceGrpc.InterNodeRpcServiceStub stub =
         connectionManager.getClientStubForHost(remoteHost);
     final StreamObserver<FlowUnitMessage> dataStream = stub.publish(serverResponseStream);
-
-    perHostOpenDataStreamMap.put(remoteHost, dataStream);
+    if (perHostOpenDataStreamMap.get(remoteHost) == null) {
+      // happens-before: updating AtomicReference.
```

Comment:
Is this comment relevant ?

Reply:
I can remove it after code review. I thought it will help in making clear how visibility is updated during the review.